8345423: Shenandoah: Parallelize concurrent cleanup

pengxiaolong opened the pull request:
Concurrent cleanup after Shenandoah collection cycle is executed by single thread(Shenandoah control thread), since currently recycling trashed regions requires heap lock even it can be done w/o heap lock. This PR is a proposal to parallelize the execution of Shenandoah concurrent cleanup after making recycling trashed regions lock free. 

With the change the time execute Concurrent cleanup has been significantly improved by 10+ times, throughput/allocation rate is also improved significantly:

TIP:
```
[30.380s][info][gc] GC(1245) Concurrent cleanup (Young) 3491M->739M(4096M) 3.634ms
[30.404s][info][gc] GC(1246) Concurrent cleanup (Young) 3258M->377M(4096M) 2.233ms
[30.434s][info][gc] GC(1247) Concurrent cleanup (Young) 2887M->333M(4096M) 7.958ms
[30.464s][info][gc] GC(1248) Concurrent cleanup (Young) 3134M->472M(4096M) 6.097ms
[30.487s][info][gc] GC(1249) Concurrent cleanup (Young) 2922M->212M(4096M) 3.072ms
[30.519s][info][gc] GC(1250) Concurrent cleanup (Young) 3404M->549M(4096M) 3.730ms
[30.552s][info][gc] GC(1251) Concurrent cleanup (Young) 3542M->712M(4096M) 6.118ms
[30.579s][info][gc] GC(1252) Concurrent cleanup (Young) 3257M->373M(4096M) 5.049ms
[30.608s][info][gc] GC(1253) Concurrent cleanup (Young) 3390M->418M(4096M) 2.779ms
```
Parallelized:
```
[30.426s][info][gc] GC(1557) Concurrent cleanup (Young) 3208M->43M(4096M) 0.177ms
[30.510s][info][gc] GC(1560) Concurrent cleanup (Young) 2938M->161M(4096M) 0.220ms
[30.534s][info][gc] GC(1561) Concurrent cleanup (Young) 2960M->57M(4096M) 0.164ms
[30.564s][info][gc] GC(1562) Concurrent cleanup (Young) 3189M->106M(4096M) 0.176ms
[30.595s][info][gc] GC(1563) Concurrent cleanup (Young) 3389M->367M(4096M) 0.247ms
[30.625s][info][gc] GC(1564) Concurrent cleanup (Young) 3662M->628M(4096M) 0.246ms
[30.649s][info][gc] GC(1565) Concurrent cleanup (Young) 3190M->150M(4096M) 0.172ms
[30.678s][info][gc] GC(1566) Concurrent cleanup (Young) 3225M->69M(4096M) 0.175ms
[30.709s][info][gc] GC(1567) Concurrent cleanup (Young) 3250M->107M(4096M) 0.179ms
[30.765s][info][gc] GC(1570) Concurrent cleanup (Young) 2932M->211M(4096M) 0.422ms
```

JVM args for the tests: `-Xms4G -Xmx4G -XX:+AlwaysPreTouch -XX:+UseShenandoahGC -XX:+UnlockExperimentalVMOptions -XX:ShenandoahGCMode=generational -XX:-ShenandoahPacing -XX:+UseTLAB -Xlog:gc`


For the same test test, but with large heap with 32G memory, the improvement on concurrent cleanup is much smaller, which might be related to less race and contention with mutator threads when the heap size is large enough:

Tip with 32G heap:
```
[5.819s][info][gc] GC(0) Concurrent cleanup (Young) 8400M->384M(32768M) 0.115ms
[6.006s][info][gc] GC(1) Concurrent cleanup (Young) 8607M->399M(32768M) 0.019ms
[6.118s][info][gc] GC(2) Concurrent cleanup (Young) 8799M->575M(32768M) 0.016ms
[6.226s][info][gc] GC(3) Concurrent cleanup (Young) 8975M->591M(32768M) 0.076ms
[6.330s][info][gc] GC(4) Concurrent cleanup (Young) 9151M->799M(32768M) 0.018ms
[9.580s][info][gc] GC(5) Concurrent cleanup (Young) 28464M->768M(32768M) 0.127ms
[14.714s][info][gc] GC(6) Concurrent cleanup (Young) 28064M->384M(32768M) 0.780ms
[20.761s][info][gc] GC(7) Concurrent cleanup (Young) 28048M->384M(32768M) 0.170ms
[26.835s][info][gc] GC(8) Concurrent cleanup (Young) 28288M->576M(32768M) 0.054ms
[32.514s][info][gc] GC(9) Concurrent cleanup (Young) 28080M->400M(32768M) 0.059ms
```
Optimized version with 32G heap:
```
[5.794s][info][gc] GC(0) Concurrent cleanup (Young) 8384M->416M(32768M) 0.022ms
[6.056s][info][gc] GC(1) Concurrent cleanup (Young) 8591M->415M(32768M) 0.032ms
[6.181s][info][gc] GC(2) Concurrent cleanup (Young) 8751M->639M(32768M) 0.026ms
[6.305s][info][gc] GC(3) Concurrent cleanup (Young) 8879M->655M(32768M) 0.020ms
[6.419s][info][gc] GC(4) Concurrent cleanup (Young) 9071M->655M(32768M) 0.052ms
[10.271s][info][gc] GC(5) Concurrent cleanup (Young) 28384M->704M(32768M) 0.269ms
[15.502s][info][gc] GC(6) Concurrent cleanup (Young) 28304M->672M(32768M) 0.087ms
[20.828s][info][gc] GC(7) Concurrent cleanup (Young) 28128M->480M(32768M) 0.061ms
[25.368s][info][gc] GC(8) Concurrent cleanup (Young) 28048M->416M(32768M) 0.139ms
[30.172s][info][gc] GC(9) Concurrent cleanup (Young) 28080M->416M(32768M) 0.055ms
```

There should be much more improvement if when there is more heap regions, especially more than 4096 regions:
Tip with 32G 4096 regions:
```
[11.015s][info][gc] GC(5) Concurrent cleanup (Young) 28115M->431M(32768M) 0.935ms
[16.771s][info][gc] GC(6) Concurrent cleanup (Young) 28151M->416M(32768M) 0.270ms
[22.709s][info][gc] GC(7) Concurrent cleanup (Young) 28136M->392M(32768M) 0.953ms
[27.757s][info][gc] GC(8) Concurrent cleanup (Young) 28120M->352M(32768M) 0.533ms
[33.029s][info][gc] GC(9) Concurrent cleanup (Young) 28096M->368M(32768M) 0.433ms
```

VS
```
[10.647s][info][gc] GC(5) Concurrent cleanup (Young) 28192M->448M(32768M) 0.174ms
[16.461s][info][gc] GC(6) Concurrent cleanup (Young) 28152M->432M(32768M) 0.286ms
[21.594s][info][gc] GC(7) Concurrent cleanup (Young) 28096M->344M(32768M) 0.334ms
[27.072s][info][gc] GC(8) Concurrent cleanup (Young) 27992M->264M(32768M) 0.283ms
[33.149s][info][gc] GC(9) Concurrent cleanup (Young) 28088M->328M(32768M) 0.365ms
```

Tip with 32G 8192 regions:
```
[7.239s][info][gc] GC(4) Concurrent cleanup (Young) 8324M->221M(32768M) 5.408ms
[11.634s][info][gc] GC(5) Concurrent cleanup (Young) 27392M->294M(32768M) 35.515ms
[17.760s][info][gc] GC(6) Concurrent cleanup (Young) 27389M->311M(32768M) 42.620ms
[24.492s][info][gc] GC(7) Concurrent cleanup (Young) 27401M->367M(32768M) 54.493ms
[30.929s][info][gc] GC(8) Concurrent cleanup (Young) 27432M->353M(32768M) 44.890ms
```
VS
```
[7.585s][info][gc] GC(4) Concurrent cleanup (Young) 8342M->160M(32768M) 0.414ms
[12.148s][info][gc] GC(5) Concurrent cleanup (Young) 27478M->295M(32768M) 0.746ms
[18.525s][info][gc] GC(6) Concurrent cleanup (Young) 27393M->229M(32768M) 1.215ms
[24.242s][info][gc] GC(7) Concurrent cleanup (Young) 27394M->220M(32768M) 0.558ms
[30.159s][info][gc] GC(8) Concurrent cleanup (Young) 27382M->214M(32768M) 1.168ms
```


### Additional test
- [x] MacOS AArch64 server fastdebug, hotspot_gc_shenandoah
- [x] Benchmarks internally setup for Ghenshen/Shenandoah

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345423](https://bugs.openjdk.org/browse/JDK-8345423): Shenandoah: Parallelize concurrent cleanup (**Enhancement** - P4)


### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer) Review applies to [4507656e](https://git.openjdk.org/jdk/pull/22538/files/4507656ee585a30dfa908241c588f176af58d238)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22538/head:pull/22538` \
`$ git checkout pull/22538`

Update a local copy of the PR: \
`$ git checkout pull/22538` \
`$ git pull https://git.openjdk.org/jdk.git pull/22538/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22538`

View PR using the GUI difftool: \
`$ git pr show -t 22538`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22538.diff">https://git.openjdk.org/jdk/pull/22538.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22538#issuecomment-2517878428)
</details>

pengxiaolong commented:
/issue add JDK-8345423

pengxiaolong commented:
/issue add JDK-8345423

pengxiaolong commented:
> Changes look good. Left some nits.

Thanks for looking into it, I'll fix them and update the PR.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -569,27 +569,61 @@ ShenandoahHeapRegion* ShenandoahHeapRegion::humongous_start_region() const {
+void ShenandoahHeapRegion::recycle_under_lock() {
+  shenandoah_assert_heaplocked();
+  if (is_trash() && _recycling.try_set()) {
+    if (is_trash()) {
```
This method is only called by mutators with heap lock, is_trash is not tested before calling the method, it is might be worth to test before calling _recycling.try_set(), otherwise mutator will mostly(fast path) behaves like:
1. _recycling.try_set() -> true (always try to perform CAS, it is slower, we want to void it in fast path).
2. is_trash() -> false and skip the recycling. 

But we want to fast path for mutator to be like: `is_trash() -> false && _recycling.is_set() -> false`.


I have remove the `is_trash` from the code path executed by concurrent cleanup in the new version, that one is not needed since `is_trash` is tested in `ShenandoahRecycleTrashedRegionsTask`

earthling-amzn replied:
Okay, I get it. The second test on line 593 is necessary because the gc workers don't hold the lock and could _in theory_ recycle the region between the first `is_trash` check on 592 and the `_recycling.try_set`.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:
```diff
@@ -882,43 +882,47 @@ size_t ShenandoahGeneration::increment_affiliated_region_count() {
 size_t ShenandoahGeneration::decrement_affiliated_region_count() {
-  shenandoah_assert_heaplocked_or_safepoint();
+  // Assertions only hold true for Java threads since they call this method under heap lock.
+  bool const is_java_thread = Thread::current()->is_Java_thread();
```
It will be weird if I only change   `decrement_affiliated_region_count` to `decrement_affiliated_region_count_under_lock` in this file, while all the other `decrement_x` / `decrease_x` /`increment_x` / `increase_x` methods in files follow the same convention.

It is probably better to add new one like `decrement_affiliated_region_count_without_lock` and not change the existing methods's behavior.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -569,27 +570,65 @@ ShenandoahHeapRegion* ShenandoahHeapRegion::humongous_start_region() const {
+    }
+    _recycling.unset();
+  } else {
+    while (_recycling.is_set()) {
```
Why are adding this? Won't this make the calling worker thread wait on another worker to recycle the region?

pengxiaolong replied:
hmm,  didn't include this intentionally, forgot to remove it from commit, sorry I'll remove it, thanks for catching it.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp:
```diff
@@ -376,7 +382,9 @@ class ShenandoahHeapRegion {
   void print_on(outputStream* st) const;
 
-  void recycle();
+  void recycle_under_lock();
```
Should be `try_recycle_under_lock` for consistency.

pengxiaolong replied:
Thanks! I have renamed it for consistency.

earthling-amzn reviewed (approved):
Thanks for the updates. It looks good to me!

pengxiaolong commented:
/reviewer 2 kdnilsen shipilev

pengxiaolong commented:
/reviewers 2 kdnilsen shipilev

pengxiaolong commented:
@kdnilsen @ysramakrishna @shipilev 

I'm gonna need more reviews for the change, thank you!

ysramakrishna reviewed (approved):
Good improvement. 

Just some minor comments.

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -88,19 +88,20 @@ ShenandoahHeapRegion::ShenandoahHeapRegion(HeapWord* start, size_t index, bool c
   if (ZapUnusedHeapArea && committed) {
     SpaceMangler::mangle_region(MemRegion(_bottom, _end));
   }
+  _recycling.unset();
```
There could be race condition that other caller immoderately set the flag, hence the assert may fail, notice similar race condition in test, that is why the double check for is_trash() was added.

ysramakrishna replied:
> There could be race condition that other caller immoderately set the flag, hence the assert may fail, 

Which other caller? We are asserting here in the constructor of the SHR object. Is this object visible to anyone other than the constructing thread at this point? I am not sure I understand the reason for a race here.

It's possible I am missing something in the lifecycle of the SHR object here.

If so, it would be good to add a brief comment on why this needs to occur here despite the constructor for the `_recycling` flag which should have been called by this point.

> notice similar race condition in test, that is why the double check for is_trash() was added.

Yes, I understood that race, which is between multiple threads potentially racing to recycle a trashed region, and resolves such a race in favor of the thread that manages to CAS  true into `_recycling` with interlocking checks for its `trash`ness.

pengxiaolong replied:
> Which other caller? 

Sorry for the confusion, not the caller of this specific method.

I meant to say the the mutator thread, mutator may call try_recycle_trashed here https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp#L1005 under heap-lock, we have removed the heap-lock from GC concurrent cleanup, therefore it becomes race condition between mutator and GC threads.

ysramakrishna replied:
OK, that's what I meant. I didn't think the region would be visible to any other thread in the time that the constructor is being executed (which, my guess was, would be when the ShenandoahHeap is first initialized -- before any mutators exist that can access the heap), but I may be missing something here in the lifecycle of a region. Thanks for pointing out the possibility of a race (but that makes me wonder about other things that may go wrong if there were such a race.) I'll think more about this later.

In any case, what I was pointing out (based on my mental model) was not a correctness issue, so I'll go away and try and understand the race you mention.

No change is needed here. It's good as is, and my review approval stands. Thanks Xiaolong!

pengxiaolong commented on src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:
```diff
@@ -882,43 +882,45 @@ size_t ShenandoahGeneration::increment_affiliated_region_count() {
   // During full gc, multiple GC worker threads may change region affiliations without a lock.  No lock is enforced
   // on read and write of _affiliated_region_count.  At the end of full gc, a single thread overwrites the count with
   // a coherent value.
```
Yes It is atomic, the lock/safepoint seems not needed. I'll probably keep the comment  as it is in this PR, since the are called from different places of FullGC and concurrentGC, we can cleanup these methods later I think.

ysramakrishna replied:
ok for now. Will be worth examining the uses from full at some point but this was just a comment so ok for now. May be leave a `TODO` comment to track if you feel like.

```
// TODO: Check and correct comment, if obsolete.
```

pengxiaolong replied:
Thanks, I'll add it if I amend any new change to this PR.

pengxiaolong commented:
> Good improvement.
> 
> Just some minor comments.

Thank you @ysramakrishna,  I have addressed all the comments except the one for comment on decrement_/increment_ methods.

ysramakrishna reviewed (approved):
Reviewed and left a couple of comments. No need for a re-review, since neither of my comments is a correctness issue that necessarily needs any code changes.

kdnilsen reviewed (approved):
Thanks.  This looks like a nice improvement.

pengxiaolong commented:
Thanks all for the reviews!

/integrate

ysramakrishna commented:
/sponsor
